jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenLayerConfigurations.java

```diff
@@ -46,85 +43,74 @@
+
+    // Gets the classes files in the 'classes' output directory.
+    Predicate<Path> isClassFile = path -> path.toString().endsWith(".class");
+    addFilesToLayer(
+        classesOutputDirectory, isClassFile, classesExtractionPath, layerBuilder::addClassFile);
+
+    // Gets the resources files in the 'classes' output directory.
+    addFilesToLayer(
+        classesOutputDirectory,
+        isClassFile.negate(),
+        resourcesExtractionPath,
+        layerBuilder::addResourceFile);
 
     // Adds all the extra files.
     if (Files.exists(extraDirectory)) {
-      try (Stream<Path> extraFilesLayerDirectoryFiles = Files.list(extraDirectory)) {
-        extraFiles = extraFilesLayerDirectoryFiles.collect(Collectors.toList());
-
-      } catch (IOException ex) {
-        throw new IOException("Failed to list directory for extra files: " + extraDirectory, ex);
-      }
+      AbsoluteUnixPath extractionBase = AbsoluteUnixPath.get("/");
+      addFilesToLayer(extraDirectory, path -> true, extractionBase, layerBuilder::addExtraFile);
     }
 
-    AbsoluteUnixPath dependenciesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_DEPENDENCIES_PATH_ON_IMAGE);
-    AbsoluteUnixPath resourcesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_RESOURCES_PATH_ON_IMAGE);
-    AbsoluteUnixPath classesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_CLASSES_PATH_ON_IMAGE);
-
-    Builder layerBuilder = JavaLayerConfigurations.builder();
-    for (Path file : dependenciesFiles) {
-      layerBuilder.addDependencyFile(file, dependenciesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : snapshotDependenciesFiles) {
-      layerBuilder.addSnapshotDependencyFile(
-          file, dependenciesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : resourcesFiles) {
-      layerBuilder.addResourceFile(file, resourcesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : classesFiles) {
-      layerBuilder.addClassFile(file, classesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : extraFiles) {
-      layerBuilder.addExtraFile(file, AbsoluteUnixPath.get("/").resolve(file.getFileName()));
-    }
     return layerBuilder.build();
   }
 
+  @FunctionalInterface
+  private static interface FileToLayerAdder {
+
+    void add(Path sourcePath, AbsoluteUnixPath pathInContainer) throws IOException;
+  }
+
+  private static void addFilesToLayer(
+      Path sourceRoot,
+      Predicate<Path> pathFilter,
+      AbsoluteUnixPath basePathInContainer,
+      FileToLayerAdder addFileToLayer)
+      throws IOException {
+    Predicate<Path> isEmptyDirectory =
+        path -> Files.isDirectory(path) && path.toFile().list().length == 0;
+    // Always add empty directories. However, ignore non-empty directories because otherwise
```

Comment:
I think we can just skip adding any directories since directories are all added in `ReproducibleLayerBuilder`.

Reply:
But empty directories won't be added, will they? Such a directory doesn't have any file, so if we don't add it, `ReproducibleLayerBuilder` won't even know it existed. Isn't it?

Reply:
Hmm, that's true - we probably want to honor the exact directories in the source files. However, I believe when you compile with Maven, it actually ignores empty directories.

Reply:
This is a general method, and we should honor the exact directory structure for other layers like extra files and resources. The program in an image may make assumptions about such empty directories.